Allow unquoted symbols in new SMT backend output if they observe the `simple_symbols` definition.

NlightNFotis opened the pull request:
Allow for symbols to be printed in an unquoted form
in the output of the new SMT backend if said symbols
are obeying the definition of a `simple_symbol` from the SMTLIB
specification.

This allows our new SMT backend to work with some solvers
that are not 100% SMTLIB compliant (namely cvc4/cvc5).

This is for now a draft, to allow CI to run and let me
observe results from test runs on CI.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis committed:
Do not quote identifiers matching definition of simple_symbol from SMT-lib spec

NlightNFotis committed:
Change regex_match to match against the identifier rather than the converted string

NlightNFotis committed:
Change tests to reflect new reality around quoting symbols.

NlightNFotis committed:
Change regression testing suite to support building with cvc4 as well as z3

NlightNFotis committed:
Fix tests to admit unquoted symbols and change solver to cvc5 for CI compatibility.

thomasspriggs commented on regression/cbmc-incr-smt2/CMakeLists.txt:
```diff
@@ -5,6 +5,16 @@ else()
+add_test_pl_profile(
+    "cbmc-new-smt-backend-cvc5"
+    "$<TARGET_FILE:cbmc> --incremental-smt2-solver 'cvc5 --lang smt --incremental' --validate-goto-model --validate-ssa-equation --slice-formula ${exclude_win_broken_tests}"
+    "-C;-X;broken-cvc5;-s;new-smt-cvc5"
```
:thinking: The `broken-cvc5` tag probably isn't needed if you have fixed all the problems locally before raising the PR.

martin-cs reviewed (commented):
Given that `cvc5` is under active and rapid development, is coming up to a major release and takes a pretty strict line on SMT-LIB compliance... is this the best way to address this problem?  Is there an issue open on the cvc5 tracker for what the problem with compliance is?

thomasspriggs commented:
> Given that `cvc5` is under active and rapid development, is coming up to a major release and takes a pretty strict line on SMT-LIB compliance... is this the best way to address this problem? Is there an issue open on the cvc5 tracker for what the problem with compliance is?

We haven't opened an issue on the cvc5 tracker. We have observed the issue this PR is intending to address on both the stable cvc4 and on cvc5. The output from cbmc is smtlib2.6 compatible both with and without the changes in this PR. I think the issue is with cvc5 rather than cbmc. Section 3.1 of the standard states the following -

>  For additional flexibility, arbitrary
sequences of whitespace and printable characters (except for | and \) enclosed in vertical bars are also allowed as symbols. Following Common Lisp’s conventions, enclosing
a simple symbol in vertical bars does not produce a new symbol. This means for instance that
abc and |abc| are the same symbol.

It seems that if we add `|` around function identifiers from the core theory such `=` to make `|=|`, the two representations of the same symbol are treated differently by cvc.

thomasspriggs committed:
Optimise check for simple symbols

NlightNFotis committed:
Change `Makefile` in `regression/cbmc-incr-smt2` to support testing with multiple solvers.

Previously the command that was enabling the new SMT backend (and providing
the script for invoking the appropriate SMT solver) was passed as part of the
options passed in `test.desc`.

Since this was removed, and passed as part of the test configuration in the
`CMakeLists.txt`, we need to do the same here.

NlightNFotis committed:
Fix bug in setting up cvc5 binary

NlightNFotis committed:
Change cmakefile to only run z3 tests on Windows

NlightNFotis commented:
Closing this, as #6777 is a substitute for this PR.
